## Problem #142: Perfect Square Collection
[Link to Problem](https://projecteuler.net/problem=142)

### Problem Description

Find the smallest $x + y + z$ with integers $x \gt y \gt z \gt 0$ such that $x + y$, $x - y$, $x + z$, $x - z$, $y + z$, $y - z$ are all perfect squares.

### Approach

Bruteforce? It appears that when trying all triples $(x, y, z)$ with values under $limit$, no triples are found for $limit = 1100$, and when trying higher limits the execution time surpasses 1 minute (_the one minute rule_) 

In [7]:
limit = 1100

def is_square(x: int) -> bool:
      return int(x ** 0.5) ** 2 == x

for x in range(1, limit + 1):
    for y in range(1, x):
            for z in range(1, y):
                if is_square(x - y) and is_square(x - z) and is_square(y - z) and is_square(x + y) and is_square(x + z) and is_square(y + z):
                    print(x, y, z, x + y + z)


We can prove that $(x, y)$ are part of a Pythagorean triple:

$$

x + y = p^2 \\[5pt]
x - y = q^2 \\[5pt]
(p + q)^2 = p^2 + 2pq + q^2 \\[5pt]
2pq = \sqrt{(x + y)(x - y)} = \sqrt{x^2 - y^2} \\[5pt]
\Rightarrow x^2 - y^2 = n^2, \; n \text{ integer} \\[5pt]
\Rightarrow n^2 + y^2 = x^2

$$

Following the same process, we can also prove that $(x, z)$ and $(y, z)$ are part of a Pythagorean triple. Therefore we can find all the Pythagorean triples up until a limit, add "edges" between the first two numbers from the triple and search for a cycle of size 3 in the constructed graph.

**PS:** Euclid's formula for any Pythagorean triple:

$$

\text{Let } x^2 = y^2 + a^2 \\[5pt]
a = m^2 - n^2 \\[5pt]
y = 2mn \\[5pt]
x = m^2 + n^2

$$

In [ ]:
def search_triples(limit):
    links = [set() for _ in range(limit + 1)]

    link_count = 0

    for m in range(1, int(limit ** 0.5) + 1):
        # Make sure n <= m and x <= limit
        for n in range(1, min(int((limit - m * m) ** 0.5) + 1, m)):
            a = m * m - n * n
            y = 2 * m * n
            x = m * m + n * n
            if x * x == y * y + a * a:
                # x + y and x - y will always be squares
                links[x].add(y)
                link_count += 1

    print("Number of links:", link_count)

    for x in range(limit + 1):
        for y in links[x]:
            for z in links[y]:
                if z in links[x]:
                    print(f"Found triple: ({x}, {y}, {z}), x + y + z = {x + y + z}")
                    return

search_triples(1000000)

Number of links: 391840
Found triple: (434657, 420968, 150568), x + y + z = 1006193


###### Result: **1006193** | Execution time: 1s

### Complexity analysis

Time complexity: $O(limit)$ (see [problem 139](../101%20-%20150/139.ipynb) for a more detailed explaination)

##### Tags: #factorials